src/tactic/positivity.lean

```diff
@@ -0,0 +1,360 @@
+      | `(%%hi ≥ %%lo) := pure (lo, hi, ff)
+      | `(%%lo < %%hi) := pure (lo, hi, tt)
+      | `(%%hi > %%lo) := pure (lo, hi, tt)
+      | _ := failed
+      end,
+      is_def_eq e hi,
+      (lo', p₂) ← norm_num.derive lo <|> refl_conv lo,
+      typ ← infer_type lo',
+      ic ← mk_instance_cache typ,
+      if strict then do
+        (ic, p₁) ← norm_num.prove_nonneg ic lo',
+        p ← mk_app ``lt_of_le_of_eq_of_lt [p₂, p₁, p₃],
+        pure (positive p)
+      else do
+        z ← mk_mapp ``has_zero.zero [some typ, none], -- there was a way to get from instance cache?
+        if lo' = z then do
+          p ← mk_app ``le_of_eq_of_le'' [p₂, p₃],
+          pure (nonnegative p)
+        else do
+          (ic, p₁) ← norm_num.prove_pos ic lo',
+          p ← mk_app ``lt_of_lt_of_eq_of_le [p₂, p₁, p₃],
+          pure (positive p)) failed
+
+/-- Attribute allowing a user to tag a tactic as an extension for the `positivity` tactic.  The main
+(recursive) step of the `positivity` tactic is to try successively all the `positivity` extensions
+on the goal, and also to try `tactic.positivity.base` on the goal. -/
+@[user_attribute]
+meta def attr : user_attribute (expr → tactic strictness) unit :=
+{ name      := `positivity,
+  descr     := "extensions handling particular operations for the `positivity` tactic",
+  cache_cfg :=
+  { mk_cache := λ ns, do
+    { t ← ns.mfoldl
+        (λ (t : expr → tactic strictness) n, do
+          t' ← eval_expr (expr → tactic strictness) (expr.const n []),
+          pure (λ e, orelse' (t' e) (t e)))
+        (λ _, failed),
+      pure (λ e, orelse' (base e) (t e)) },
+    dependencies := [] } }
+
+/-- Look for a proof of positivity/nonnegativity of an expression `e`; if found, return the proof
+together with a `strictness` stating whether the proof found was for strict positivity
+(`positive p`) or only for nonnegativity (`nonnegative p`). -/
+meta def core (e : expr) : tactic strictness := do
+  f ← attr.get_cache,
+  f e <|> fail "failed to prove positivity/nonnegativity"
+
+end positivity
+
+namespace interactive
+
+setup_tactic_parser
+
+/-- Tactic solving goals of the form `0 ≤ x` and `0 < x`.  The tactic works recursively according to
+the syntax of the expression `x`, if the atoms composing the expression all have numeric lower
+bounds which can be proved positive/nonnegative by `norm_num`.  This tactic either closes the goal
+or fails. -/
+meta def positivity : tactic unit := focus1 $ do
+  t ← target,
+  (rel_desired, a) ← match t with
+  | `(0 ≤ %%e₂) := pure (ff, e₂)
+  | `(%%e₂ ≥ 0) := pure (ff, e₂)
+  | `(0 < %%e₂) := pure (tt, e₂)
+  | `(%%e₂ > 0) := pure (tt, e₂)
+  | _ := fail "not a positivity/nonnegativity goal"
+  end,
+  strictness_proved ← tactic.positivity.core a,
+  match rel_desired, strictness_proved with
+  | tt, (positive p) := pure p
+  | tt, (nonnegative _) := fail ("failed to prove strict positivity, but it would be possible to" ++
```

Comment:
```suggestion
  | tt, (nonnegative _) := fail ("failed to prove strict positivity, but it would be possible to " ++
```
since `toprove` isn't a word!